### PR TITLE
executor: handle the corner case that temp index is not exist but the normal index is exist 

### LINKED ISSUE #51784: insert ignore may cause data and index inconsistence
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

``` SQL
store, dom := testkit.CreateMockStoreAndDomain(t)

	tk := testkit.NewTestKit(t, store)
	tk.MustExec("use test")
	tk.MustExec("create table t(a smallint(6) DEFAULT '-13202', b varchar(221) NOT NULL DEFAULT '9k23s0uioddlcahb5qmd9mulds8hd6go0q6cxlsshbj3w9y08yuqcky1gv90eldi4ylbno8ofs15o2exlyka6vt653idyy84b0s8qxtnxq501', " +
		"c tinyint(1) NOT NULL DEFAULT '0', PRIMARY KEY (c, b));")

	tk1 := testkit.NewTestKit(t, store)
	tk1.MustExec("use test")

	tk3 := testkit.NewTestKit(t, store)
	tk3.MustExec("use test")

	d := dom.DDL()
	originalCallback := d.GetHook()
	defer d.SetHook(originalCallback)
	callback := &callback.TestDDLCallback{}
	//_, err := tk1.Exec("INSERT INTO t VALUES (-18585,'+zmNHMov7KF~tF*DFb',1), (-18585,'0',1), (-18585,'1',1), (-18585,'9k23s0uioddlcahb5qmd9mulds8hd6go0q6cxlsshbj3w9y08yuqcky1gv90eldi4ylbno8ofs15o2exlyka6vt653idyy84b0s8qxtnxq501',1), (-18585,'Q',1), (-18585,'s',1);;")
	//assert.NoError(t, err)

	onJobUpdatedExportedFunc := func(job *model.Job) {
		if t.Failed() {
			return
		}
		switch job.SchemaState {
		case model.StateDeleteOnly:
			_, err := tk1.Exec("INSERT INTO t VALUES (-18585,'+zmNHMov7KF~tF*DFb',1), (-18585,'0',1), (-18585,'1',1), (-18585,'9k23s0uioddlcahb5qmd9mulds8hd6go0q6cxlsshbj3w9y08yuqcky1gv90eldi4ylbno8ofs15o2exlyka6vt653idyy84b0s8qxtnxq501',1), (-18585,'Q',1), (-18585,'s',1);;")
			assert.NoError(t, err)
		case model.StateWriteReorganization:
			idx := testutil.FindIdxInfo(dom, "test", "t", "idx")
			if idx.BackfillState == model.BackfillStateReadyToMerge {
				_, err := tk1.Exec("insert ignore into `t`  values ( '06j$GwD*zizDUEEy&*-',null,-2028 ) ,( '((J)',0,17786 ) ,( 'SD$ruyO$%c',1,null ) ,( 'a3Ex',0,24877 ) ,( 'O=M',1,31913 ) ,( '7njQI*',0,17610 );")
				assert.NoError(t, err)
				return
			}
		}
	}
	callback.OnJobUpdatedExported.Store(&onJobUpdatedExportedFunc)
	d.SetHook(callback)

	tk.MustExec("alter table t add unique index idx(b);")
	tk.MustExec("admin check table t;")
```

### 2. What did you expect to see? (Required)
test pass
### 3. What did you see instead (Required)
failed
### 4. What is your TiDB version? (Required)
master

### PR DESCRIPTION
**Pull request by wjhuang2016**
### What problem does this PR solve?

Issue Number: close #51784

Problem Summary:
If the temp index is not exist, we need to check if the normal index is exist.
The corner case is showed in the added unit test.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
fix a bug may cause unique index and data inconsistent when using insert ignore
```

### CODE REVIEW

**Review by Benjamin2037 — APPROVED** [review]
LGTM

File: pkg/executor/insert_common.go [hunk 1]
```diff
@@ -1256,42 +1277,44 @@ CheckAndInsert:
 			}
 		}
 
+		rowInserted := false
 		for _, uk := range r.uniqueKeys {
```
Comment (zimulala) [inline pkg/executor/insert_common.go:1281]: Could we do the logic as follows? Then we can remove line1296-1317.
```
ukNewKey := uk.newKey
if tablecodec.IsTempIndexKey(ukNewKey) {
    tablecodec.TempIndexKey2IndexKey(ukNewKey)
}
```

File: pkg/executor/insert_common.go [hunk 2]
```diff
@@ -1256,42 +1277,44 @@ CheckAndInsert:
...
-				if handle == nil {
-					continue
+				if rowInserted {
+					break
 				}
-				_, err = e.removeRow(ctx, txn, handle, r, true)
-				if err != nil {
+				continue
+			}
+			if tablecodec.IsTempIndexKey(uk.newKey) {
```
[the remark this answers is not in the document]
Comment (wjhuang2016) [reply pkg/executor/insert_common.go:1290]: uk.newKey is temp key and not exist here.

File: pkg/executor/insert_common.go (hunk 1 above)
Comment (wjhuang2016) [reply pkg/executor/insert_common.go:1275]: No, the logic changes.

Comment (zimulala) [reply pkg/executor/insert_common.go:1275]: Got it. In the case of tmpIdx, we need to check both scenarios, not just one.

File: pkg/executor/insert_common.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (zimulala) [reply pkg/executor/insert_common.go:1290]: The tmpKey is obtained in the tmpKey mode. 
If no Tmpkey is found, the tmpkey is converted to the normal key mode and obtained again.

**Review by zimulala — APPROVED** [review]

File: pkg/ddl/db_test.go [hunk 3]
```diff
@@ -1064,6 +1066,43 @@ func TestMDLTruncateTable(t *testing.T) {
...
+	callback := &callback.TestDDLCallback{}
+
+	onJobUpdatedExportedFunc := func(job *model.Job) {
+		switch job.SchemaState {
+		case model.StateDeleteOnly:
+			_, err := tk1.Exec("INSERT INTO t VALUES (-18585,'aaa',1), (-18585,'0',1), (-18585,'1',1), (-18585,'duplicatevalue',1);")
+			assert.NoError(t, err)
+		case model.StateWriteReorganization:
+			idx := testutil.FindIdxInfo(dom, "test", "t", "idx")
+			if idx.BackfillState == model.BackfillStateReadyToMerge {
```
Comment (zimulala) [inline pkg/ddl/db_test.go:1092]: This case is:
* State DO: insert data, no key
* State WR: backfill normal keys, no tmp key.
